src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
+                    std::optional<std::set<CTxDestination>> t_filter = std::set<CTxDestination>({keyId});
+                    return std::make_pair(t_filter, AddrSet::Empty());
+                },
+                [&](const CScriptID& scriptId) {
+                    std::optional<std::set<CTxDestination>> t_filter = std::set<CTxDestination>({scriptId});
+                    return std::make_pair(t_filter, AddrSet::Empty());
+                },
+                [&](const auto& other) {
+                    std::optional<std::set<CTxDestination>> t_filter = std::nullopt;
+                    return std::make_pair(t_filter, AddrSet::ForPaymentAddresses({addr}));
                 }
-                info.vjsout.push_back(jso);
-            }
-        }
-
-        // create output for any change
-        if (jsChange>0) {
-            info.vjsout.push_back(JSOutput(std::get<libzcash::SproutPaymentAddress>(frompaymentaddress_), jsChange));
+            }, addr);
+        },
+        [&](const FromAnyTaddr& taddr) {
+            std::optional<std::set<CTxDestination>> t_filter = std::set<CTxDestination>({});
+            return std::make_pair(t_filter, AddrSet::Empty());
+        }
+    }, paymentSource_);
+
+    if (filters.first.has_value()) {
+        pwalletMain->AvailableCoins(
+                unspent.utxos,
+                false,                    // fOnlyConfirmed
+                nullptr,                  // coinControl
+                true,                     // fIncludeZeroValue
+                allowTransparentCoinbase, // fIncludeCoinBase
+                true,                     // fOnlySpendable
+                mindepth_,                // nMinDepth
+                &filters.first.value());  // onlyFilterByDests
+    }
+
+    pwalletMain->GetFilteredNotes(
+            unspent.sproutNoteEntries,
+            unspent.saplingNoteEntries,
+            filters.second,
+            mindepth_);
+
+    return unspent;
+}
 
-            LogPrint("zrpcunsafe", "%s: generating note for change (amount=%s)\n",
-                    getId(),
-                    FormatMoney(jsChange)
-                    );
+bool SpendableInputs::LimitToAmount(CAmount amountRequired, CAmount dustThreshold) {
+    // select Sprout notes for spending first
+    std::sort(sproutNoteEntries.begin(), sproutNoteEntries.end(),
+        [](SproutNoteEntry i, SproutNoteEntry j) -> bool {
+            return i.note.value() > j.note.value();
+        });
+    auto sproutIt = sproutNoteEntries.begin();
+    while (sproutIt != sproutNoteEntries.end() && amountRequired > 0) {
+        amountRequired -= sproutIt->note.value();
+        ++sproutIt;
+        // if the total would result in change below the dust threshold,
+        // adjust the amountRequired so that we select additional inputs
+        if (amountRequired < 0 && amountRequired > -dustThreshold) {
+            amountRequired += dustThreshold;
+            dustThreshold = amountRequired;
+        }
```

Comment:
Consider the following example:
```
dustThreshold = 10
amountRequired = 1000
note1.value = 1001
note2.value = 10
```
After the first iteration we have `amountRemaining = dustThreshold = 9`. Then on the second iteration we subtract 10 from `amountRemaining` and hit the "change below the dust threshold" condition again and we have `amountRemaining = dustThreshold = 8`, having spent 1011. Spending 1011, with 1000 going to the recipient and 11 to change, is already sufficient because 11 is not dust. But the algorithm does not recognize that and wants to spend at least 8 more.

Reply:
Ah, right, because we don't have any explicit transition to a separate "avoiding dust" mode if I try to reuse `amountRequired` for this purpose.

Reply:
Fixed.